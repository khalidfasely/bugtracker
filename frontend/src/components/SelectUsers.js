import { useState } from 'react';
import Select from 'react-select';
import makeAnimated from 'react-select/animated';

const SelectUsers = ({ users, selectedUsers, setSelectedUsers, selectedAdmins, setSelectedAdmins, errorSelect }) => {

    let options = users?.map(user => ({ value: user, label: user }));

    const customTheme = (theme) => ({
        ...theme,
        colors: {
            ...theme.colors,
            primary25: 'orange',
            primary: 'red'
        }
    });

    const onChangeUsers = (selectedUsers, { action, removedValue }) => {
        switch (action) {
            case "remove-value":
            case "pop-value":
                if (removedValue.isFixed) {
                  return;
                }
                break;
            case "clear":
                selectedUsers = selectedUsers.filter(v => v.isFixed);
                break;
            default:
                break;
        };
        setSelectedUsers(selectedUsers);
    };

    const onChangeAdmins = (selectedAdmins, { action, removedValue }) => {
        switch (action) {
            case "remove-value":
            case "pop-value":
                if (removedValue.isFixed) {
                  return;
                }
                break;
            case "clear":
                selectedAdmins = selectedAdmins.filter(v => v.isFixed);
                break;
            default:
                break;
        };
        setSelectedAdmins(selectedAdmins);
    };

    return (
        <div>
            {errorSelect && <p>{errorSelect}</p>}
            <Select
                components={makeAnimated()}
                value={selectedUsers}
                theme={customTheme}
                defaultValue={selectedUsers}
                onChange={onChangeUsers}
                options={options}
                isClearable={!options && options.some(option => option.isFixed)}
                placeholder='Select Users'
                noOptionsMessage={() => 'No other user!'}
                isMulti
                isSearchable
            />
            <Select
                components={makeAnimated()}
                value={selectedAdmins}
                theme={customTheme}
                defaultValue={selectedAdmins}
                onChange={onChangeAdmins}
                options={selectedUsers}
                isClearable={!options && options.some(option => option.isFixed)}
                placeholder='Select Admins'
                noOptionsMessage={() => 'No other user!'}
                isMulti
                isSearchable
            />
        </div>
    );
}

export default SelectUsers;