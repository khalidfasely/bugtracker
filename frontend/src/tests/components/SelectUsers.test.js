import React from "react";
import { render } from "@testing-library/react";

import SelectUsers from "../../components/SelectUsers";

//Mock the react-select :: https://polvara.me/posts/testing-a-custom-select-with-react-testing-library
jest.mock("react-select", () => ({ options, value, onChange }) => {
    function handleChange(event) {
      const option = options.find(
        (option) => option.value === event.currentTarget.value
      );
      onChange(option);
    }
  
    return (
      <select multiple={true} data-testid="select" value={value} onChange={handleChange}>
        {options.map(({ label, value }) => (
          <option key={Math.random() * 10} value={value}>
            {label}
          </option>
        ))}
      </select>
    );
});

let props;

beforeEach(() => {
    props = {
        users: ["admin", "1"],
        selectedUsers: ["admin"],
        setSelectedUsers: jest.fn(),
        selectedAdmins: ["admin"],
        setSelectedAdmins: jest.fn(),
        errorSelect: null
    }
});

test('Should render SelectUsers components correctly', () => {
    const { asFragment } = render(<SelectUsers {...props} />);
    expect(asFragment()).toMatchSnapshot();
});