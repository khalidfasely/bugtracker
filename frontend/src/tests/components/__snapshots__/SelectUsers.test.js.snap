// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render SelectUsers components correctly 1`] = `
<DocumentFragment>
  <div>
    <select
      data-testid="select"
    >
      <option
        value="admin"
      >
        admin
      </option>
      <option
        value="1"
      >
        1
      </option>
    </select>
    <select
      data-testid="select"
    >
      <option />
    </select>
  </div>
</DocumentFragment>
`;
